### PR TITLE
Check if call return type is visibly uninhabited when building MIR

### LINKED ISSUE #93161: ICE for generators involving Never type
Building the following code ICEs with current master (not yet on nightly):
```rust
#![feature(never_type)]

use std::future::Future;

// See if we can run a basic `async fn`
pub async fn foo(x: &u32, y: u32) -> u32 {
    let y = &y;
    let z = 9;
    let z = &z;
    let y = async { *y + *z }.await;
    let a = 10;
    let a = &a;
    *x + y + *a
}

async fn add(x: u32, y: u32) -> u32 {
    let a = async { x + y };
    a.await
}

async fn build_aggregate(a: u32, b: u32, c: u32, d: u32) -> u32 {
    let x = (add(a, b).await, add(c, d).await);
    x.0 + x.1
}

enum Never {}
fn never() -> Never {
    panic!()
}

async fn includes_never(crash: bool, x: u32) -> u32 {
    let mut result = async { x * x }.await;
    if !crash {
        return result;
    }
    #[allow(unused)]
    let bad = never();
    result *= async { x + x }.await;
    drop(bad);
    result
}

async fn partial_init(x: u32) -> u32 {
    #[allow(unreachable_code)]
    let _x: (String, !) = (String::new(), return async { x + x }.await);
}

async fn read_exact(_from: &mut &[u8], _to: &mut [u8]) -> Option<()> {
    Some(())
}

async fn hello_world() {
    let data = [0u8; 1];
    let mut reader = &data[..];

    let mut marker = [0u8; 1];
    read_exact(&mut reader, &mut marker).await.unwrap();
}

fn run_fut<T>(fut: impl Future<Output = T>) -> T {
    use std::sync::Arc;
    use std::task::{Context, Poll, Wake, Waker};

    struct MyWaker;
    impl Wake for MyWaker {
        fn wake(self: Arc<Self>) {
            unimplemented!()
        }
    }

    let waker = Waker::from(Arc::new(MyWaker));
    let mut context = Context::from_waker(&waker);

    let mut pinned = Box::pin(fut);
    loop {
        match pinned.as_mut().poll(&mut context) {
            Poll::Pending => continue,
            Poll::Ready(v) => return v,
        }
    }
}

fn main() {
    let x = 5;
    assert_eq!(run_fut(foo(&x, 7)), 31);
    assert_eq!(run_fut(build_aggregate(1, 2, 3, 4)), 10);
    assert_eq!(run_fut(includes_never(false, 4)), 16);
    assert_eq!(run_fut(partial_init(4)), 8);
    run_fut(hello_world());
}
```
Output:
```
error: internal compiler error: compiler/rustc_mir_transform/src/generator.rs:755:13: Broken MIR: generator contains type Never in MIR, but typeck only knows about {ResumeTy, bool, u32, impl Future<Output = [async output]>, (), impl Future<Output = [async output]>} and [bool, u32]
  --> tests/run-pass/async-fn.rs:31:53
   |
31 |   async fn includes_never(crash: bool, x: u32) -> u32 {
   |  _____________________________________________________^
32 | |     let mut result = async { x * x }.await;
33 | |     if !crash {
34 | |         return result;
...  |
40 | |     result
41 | | }
   | |_^

thread 'rustc' panicked at 'Box<dyn Any>', /rustc/84e918971d643c6a33067d5125214ab800ce5307/compiler/rustc_errors/src/lib.rs:1115:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.60.0-nightly (84e918971 2022-01-21) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [optimized_mir] optimizing MIR for `includes_never::{closure#0}`
#1 [layout_of] computing layout of `[static generator@tests/run-pass/async-fn.rs:31:53: 41:2]`
end of query stack

```

This is a regression, the same code worked fine yesterday. Likely cause: https://github.com/rust-lang/rust/pull/91032

### PR DESCRIPTION
**Pull request by tmiasko**
The main motivation behind the change is to expose information about diverging
calls to the generator transform and match the precision of drop range tracking
which already understands that call expressions with visibly uninhabited types
diverges.

This change should also accept strictly more programs than before. That is
programs that were previously rejected due to errors raised by control-flow
sensitive checks in a code that is no longer considered reachable.

Fixes #93161.

### CODE REVIEW

File: compiler/rustc_mir_build/src/build/expr/into.rs [hunk 1]
```diff
@@ -255,10 +255,11 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                         func: fun,
                         args,
                         cleanup: None,
-                        // FIXME(varkor): replace this with an uninhabitedness-based check.
-                        // This requires getting access to the current module to call
-                        // `tcx.is_ty_uninhabited_from`, which is currently tricky to do.
-                        destination: if expr.ty.is_never() {
+                        destination: if this.tcx.is_ty_uninhabited_from(
+                            this.parent_module,
```
Comment (Aaron1011) [inline compiler/rustc_mir_build/src/build/expr/into.rs:259]: I think this is overly conservative - if the return type is uninhabited (even if the uninhabitedness is due to a type from another crate), then it should always be sound to omit the destination.

Comment (eholk) [reply compiler/rustc_mir_build/src/build/expr/into.rs:259]: Is there a better way to check uninhabitedness? I've written some code that depends on this too, so if there's a better check I'd love to use it there too.

Comment (tmiasko) [reply compiler/rustc_mir_build/src/build/expr/into.rs:259]: The rationale for taking visibility into account is to ensure that uninhabitedness remains an implementation detail unless it is visible to code in question. For example why this matters see a new test case privately-uninhabited-mir-call.rs.

Comment (Aaron1011) [reply compiler/rustc_mir_build/src/build/expr/into.rs:259]: Oh, that's interesting - I didn't realize that this would affect what code we accept. In that case, taking visibility into account sounds correct. Could you add a comment explaining that?

File: compiler/rustc_session/src/options.rs [hunk 2]
```diff
@@ -1166,6 +1166,8 @@ options! {
     dont_buffer_diagnostics: bool = (false, parse_bool, [UNTRACKED],
         "emit diagnostics rather than buffering (breaks NLL error downgrading, sorting) \
         (default: no)"),
+    drop_tracking: bool = (false, parse_bool, [TRACKED],
```
Comment (eholk) [inline compiler/rustc_session/src/options.rs:1169]: I might call this flag `generator_drop_tracking` to make the scope more clear, although I think it's fine if you want to leave it as is since the help text makes the scope clear.

File: src/test/ui/generator/issue-93161.rs [hunk 3]
```diff
@@ -32,7 +32,7 @@ fn never() -> Never {
 }
 
 async fn includes_never(crash: bool, x: u32) -> u32 {
-    let mut result = async { x * x }.await;
+    let result = async { x * x }.await;
```
[the remark this answers is not in the document]
Comment (pnkfelix) [reply src/test/ui/generator/issue-93161.rs:36]: To be clear: Wesley's assumption is that the unused `mut` lint started firing here (because the mutation of `result` falls after the call to `never()` below, and so that's why the PR author has removed the `mut`.

And Wesley's question is: Do we need to worry about this? Considering e.g. code that might have `#![deny(unused_mut)]`, and now it starts seeing that as an error.

(My personal answer is that lints are allowed to change their behavior in this manner. I'm currently double-checking/asking around for examples from Rust's past.)

[the remark this answers is not in the document]
Comment (wesleywiser) [reply src/test/ui/generator/issue-93161.rs:36]: Additional discussion in https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/lints.2C.20flow-sensitivity.2C.20and.20stability

[the remark this answers is not in the document]
Comment (tmiasko) [reply src/test/ui/generator/issue-93161.rs:36]: The lint side of things seems fine to me from compatibility perspective. It is
not unusual for us to adjust how lints work, and practical impact is limited by
cargo capping the lint level for dependencies.

The fact that we would now accept more programs seems more significant, since
it would be a breaking change to revert. Though, as far as I can see, we did in
fact accept those programs after #54125 landed, but have since regressed after
switching to MIR borrowck.

**Comment by eholk** [comment]
What's the status on this PR? Is it still waiting on a review, or are there changes that still need to be made?

**Comment by tmiasko** [comment]
It is waiting for review.

(I would be happy to review the change in other direction if this blocks drop tracking. That is, reduce precision on the typeck side and consider only call expressions with never type as diverging to match MIR building).

**Comment by eholk** [comment]
Sounds good. I just opened #94752, which pulls the fix I had added out of my earlier PR. That should be a little quicker to merge than this one, and it can coexist with this change, so I think it makes sense to go ahead and merge it. Once this change lands, we can consider reverting mine, although leaving it in wouldn't hurt anything either.

**Comment by tmandry** [comment]
Looks good, but I think it's missing a test for #93161 and the case of new programs being accepted (e.g. non-privately-uninhabited version of `privately-uninhabited-mir-call.rs`)

Another thing I want to check: What if the uninhabited type is `#[non_exhaustive]`? Seems like we should be conservative in that case.

**Comment by tmandry** [comment]
Oops, sorry for missing those.

> `is_ty_uninhabited_from` considers `#[non_exhaustive]` enums from other crates to be inhabited.

Can you verify if there is a test that exercises this, and add one if not?

Otherwise looks good to me, but I'm checking with t-lang that we want to commit to this behavior.

**Comment by nikomatsakis** [comment]
Actually, I read the source and I think I understand now.

**Comment by nikomatsakis** [comment]
My take:

This is definitely a change to the language definition and hence the lang team ought to weigh in on it, so I've adjusted the labels.

**Comment by tmiasko** [comment]
In #54125, a control flow was made more precise by considering a call
expressions with visibly uninhabited type to never returned normally.

The borrow-checking and const-checking are flow sensitive, so we started to
accept programs that were previously rejected due to code that was no longer
considered to be reachable. For example:

```rust
pub enum Unimplemented {}

const fn todo() -> Unimplemented {
    panic!();
}

fn f() {
    let a: u32;
    todo();
    &a; // Previously: borrow of possibly-uninitialized variable
}

fn g() {
    let b = 10;
    todo();
    b = 15; // Previously: cannot assign twice to immutable variable
}

const fn h<T>(c: T) { // Previously: destructors cannot be evaluated at compile-time
    todo();
}
```

The implementation from #54125 was limited to analysis working on HIR, with a
MIR counterpart left out as a future work. In the meantime, those control flow
sensitive checks migrated to MIR and the behaviour regressed. This pull
requests re-implements changes from #54125 on MIR and accepts those programs
again.

The main motivation behind those changes is to avoid capturing locals from
unreachable code in generators, although that aspect is just an optimization
(on the MIR side).

**Comment by nikomatsakis** [comment]
@tmiasko thanks! That is very helpful, and I had forgotten the context that we used to do this analysis on HIR and regressed the behavior.

**Comment by nikomatsakis** [comment]
Here is my summary of what is going on (@tmiasko please correct me if I am missing details!). On the basis of this write-up, I believe we should merge this PR, so...

@rfcbot fcp merge

### The context

In Rust today, you can have functions that are guaranteed never to return by writing `-> !`:

```rust
fn never_returns() -> ! { panic!() }
```

we take this into account in various ways, including the dead-code lint but also in MIR borrow check ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=8bbce8b3a089eea0b658331262d02513) -- thanks to @tmiasko for the example):

```rust
fn main() {
    let a: u32;
    never_returns();
    &a; // No error, even though `a` is not initialized, because this code never run
}
```

**At the moment,** when you return an uninhabited type like `enum Unimplemented { }`, we don't recognize this type as uninhabited for the purposes of borrow check and other checks that are performed on MIR ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=ef934da52445016ebb8768036df4fcfc)):

```rust
enum Uninhabited { }
fn never_returns() -> Uninhabited { panic!() }
fn main() {
    let a: u32;
    never_returns();
    &a; // ERROR borrow of possibly-uninitialized variable: `a`
}
```

This is true *even though* we warn that the code is unreachable:

```
warning: unreachable expression
 [--> src/main.rs:6:6
](https://play.rust-lang.org/#)  |
5 |     never_returns();
  |     --------------- any code following this expression is unreachable
6 |     &a; // ERROR
  |      ^ unreachable expression
  |
  = note: `#[warn(unreachable_code)]` on by default
```

### Historical note

According to @tmiasko, this [was not always the case](https://github.com/rust-lang/rust/pull/93313#issuecomment-1075608856):

> In https://github.com/rust-lang/rust/pull/54125, a control flow was made more precise by considering a call
expressions with visibly uninhabited type to never returned normally.

### Proposed change

We will consider any function that returns a type `Foo` which is "visibly uninhabited" to never return. *Visibly uninhabited* means that some public field has an uninhabited type. This means that the previous example compiles, but this one doesn't:

```rust
mod secret {
    enum Uninhabited { }
    pub struct Wrapper { u: Uninhabited }
    pub fn never_returns() -> Wrapper { panic!() }
}
fn main() {
    let a: u32;
    secret::never_returns();
    &a; // ERROR
}
```

Notably, it *also* does not give a dead-code warning so it is consistent. If the field `Wrapper::u` is made `pub`, then the code does compile (and gets a warning).

### Implications

* I believe the *borrow check* is the primary flow-sensitive analysis that is affected.
* The Unsafe Check used to be flow-sensitive, but [we changed that](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=32febf137e5855f1c8998600926909b0), so I wouldn't expect it to change behavior here (@tmiasko -- this would be a nice test).
* Apparently this impacts the optimization of generators as well. =)

Anybody have further suggestions of things that change behavior?

**Comment by tmiasko** [comment]
> The Unsafe Check used to be flow-sensitive, but we changed that, so I wouldn't expect it to change behavior here (this would be a nice test).

Interesting that unsafe check was flow-sensitive. The issue was that the unreachable blocks had been removed before the check, i.e., it was accidentally flow-sensitive rather than inherently so. 

I extended the existing test to cover call with uninhabited types as well.

**Comment by nikomatsakis** [comment]
> Interesting that unsafe check was flow-sensitive. The issue was that the unreachable blocks had been removed before the check, i.e., it was accidentally flow-sensitive rather than inherently so.

Yes, this is what I meant really :)

**Comment by scottmcm** [comment]
I went through a bunch of thoughts here.  It's not obvious to me that the `mut` check should be affected, but it looks like this is [already fine](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=96cb3a22ce72d478542e1c9355c7bf9b):
```rust
    let x = 4;
    panic!();
    x *= 4;
```
So I think it's fine to have it be impacted here.

The visibly uninhabited distinction is great.  One thing that's plausibly already accounted for, but I didn't see a test for (and I forget how to look at query details) is `non_exhaustive`.  Can you please add a test that something like `#[non_exhaustive] pub enum NeverButLaterSometimes {}` is handled correctly?

@rfcbot concern confirm-non-exhaustive

But overall, I'm glad to see this.  The more we can make `!` not special compared to other uninhabited types, the better.

@rfcbot reviewed

**Comment by tmiasko** [comment]
> The visibly uninhabited distinction is great. One thing that's plausibly already accounted for, but I didn't see a test for (and I forget how to look at query details) is `non_exhaustive`. Can you please add a test that something like `#[non_exhaustive] pub enum NeverButLaterSometimes {}` is handled correctly?

@scottmcm  does https://github.com/rust-lang/rust/pull/93313#issuecomment-1071204355 address this concern, or would you like me to add a dedicated test case?

**Comment by scottmcm** [comment]
Oh, doh, I guess I only Ctrl-F'd in the diff, not the chat.

I think that's probably fine, especially since making tests for non-exhaustive is a pain.

@rfcbot resolve confirm-non-exhaustive

**Comment by eholk** [comment]
Since the FCP is passed, are there any more changes needed or can we go ahead and merge it?

@tmandry - You had an r=me earlier in the thread. Does this still look good to you?

File: src/test/ui/generator/issue-93161.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (RalfJung) [reply src/test/ui/generator/issue-93161.rs:36]: It also seems surprising that this `mut` would be unused, since `result` *is* mutated in later code -- in a regular function, we are not that smart wrt unused-mut earnings, or are we?

[the remark this answers is not in the document]
Comment (eholk) [reply src/test/ui/generator/issue-93161.rs:36]: I was talking about this with @scottmcm a little while back and he found this example: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=26220b66188903b1dd8f3325960f75af

That version compiles even though `x` is mutated, since the mutation comes after the panic. Although, for me making it `let mut x` doesn't raise an unused mut warning. Anyway, I think this shows prior art for being clever about things being mutated only in unreachable code, so this change seems to make things more uniform.

File: src/test/ui/consts/const-eval/validate_uninhabited_zsts.32bit.stderr [hunk 4]
```diff
@@ -10,14 +10,11 @@ LL |     unsafe { std::mem::transmute(()) }
...
    |                          ----- inside `FOO` at $DIR/validate_uninhabited_zsts.rs:13:26
 
-error[E0080]: it is undefined behavior to use this value
-  --> $DIR/validate_uninhabited_zsts.rs:16:1
+error[E0080]: evaluation of constant value failed
+  --> $DIR/validate_uninhabited_zsts.rs:16:35
    |
 LL | const BAR: [Empty; 3] = [unsafe { std::mem::transmute(()) }; 3];
-   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type validation failed at [0]: encountered a value of uninhabited type Empty
-   |
```
[the remark this answers is not in the document]
Comment (RalfJung) [reply src/test/ui/consts/const-eval/validate_uninhabited_zsts.32bit.stderr:18]: I did the same [in a Miri test](https://github.com/rust-lang/miri/pull/2078/files#diff-f422314ad7e7c2105673b9fe090e4798ca3d64b938602bbb6bf14c4736a2b107) so that type definition can probably just be copied.

**Comment by RalfJung** [comment]
I guess the other question is, do we really want `pub` annotations to affect the *operational behavior* of a program? Or should `pub` only affect *whether* a program compiles, but never affect what the program does when it runs (and whether it has UB)?

Having the borrow checker take into account visibility makes sense to me, but somehow this ended up changing interpreter (Miri) behavior and that is worrying.

**Comment by eholk** [comment]
@RalfJung - can you summarize how the behavior of Miri changed? Is it that what used to be a runtime error because a compiler error, or vice-versa? Or was it always a runtime error, but the error is different now?

If all that's changed is that Miri is reporting undefined behavior for different reasons, but the set of things that are UB is unchanged, that seems okay to me?

**Comment by RalfJung** [comment]
`std::mem::transmute::<(), Void>(())` (with `Void` being a variantless enum) now fails with "called transmute on uninhabited type". That's because it doesn't even get a return place any more, so it has to abort very early -- it cannot even do the usual things transmute does, because those need a return place to put the result into.

But yeah, if there was a return place validation would then immediately return UB anyway. The testcases disabled validation to be able to exhibit some other code paths in Miri, and that's why the UB location changed, but with Miri's default flags only the UB message should change.
